### PR TITLE
fix(es/hygiene): Hoist parameter in arrow and constructor

### LINKED ISSUE #7248: ReferenceError on a valid code
### Describe the bug

Bug found while migrating actual project from CRA to Vite using defaults. Valid JS code is transpired incorrectly. Reproducible on the default configuration via `npx swc` or the playground.

### Input code

```typescript
const messages = [{ message: 'hello' }];
const logs = messages.map(({ log = () => console.log(message), message }) => log);
logs[0]();
```

### Config

```json
{
  "jsc": {
    "parser": {
      "syntax": "ecmascript",
      "jsx": false
    },
    "target": "es5",
    "loose": false,
    "minify": {
      "compress": false,
      "mangle": false
    }
  },
  "module": {
    "type": "es6"
  },
  "minify": false,
  "isModule": true
}
```

### Playground link

https://play.swc.rs/?version=1.3.49&code=H4sIAAAAAAAAA0vOzysuUchNLS5OTE8tVrBViK6G8awU1DNSc3Ly1RVqY625ksEKc%2FLTQYpg6vVyEws0NKpBwkBRDU0FWzsFkML8nFQ9oJgGVJ2mDkyHQi1YDVBO05oLZFi0QawGkAkAxy084IcAAAA%3D&config=H4sIAAAAAAAAA1WPSw7DIAwF9zkF8rrbdtE79BAWdSIifrKJVBTl7iUE0maH3xsz8jooBbNoeKq1PMsQkYX4nEsi2Sf8lARIOxTNJia49XaWvRrRCtVoOxpIyBOluiX3hoMNQajjLXPGmzH%2FC3VwkUnkCu4o%2BsnSVTc0JbjwXmrZDkk50qF%2FwA%2FqsvNjMPLqm4kXGrYvhlQioBQBAAA%3D

### Expected behavior

Outputs `hello` into the console

### Actual behavior

JS error: `Uncaught ReferenceError: message is not defined`

### Version

1.3.49

### Additional context

_No response_

### PR DESCRIPTION
**Pull request by Austaras**
**Description:**

closes #7248 
**BREAKING CHANGE:**

**Related issue (if exists):**

### CODE REVIEW

File: crates/swc_ecma_transforms_base/src/resolver/mod.rs [hunk 1]
```diff
@@ -549,6 +549,18 @@ impl<'a> VisitMut for Resolver<'a> {
 
             let old = child.ident_type;
             child.ident_type = IdentType::Binding;
+            {
+                let params = e
+                    .params
+                    .iter()
+                    .filter(|p| !p.is_rest())
+                    .flat_map(find_pat_ids)
+                    .collect::<Vec<Id>>();
```
Comment (kdy1) [inline crates/swc_ecma_transforms_base/src/resolver/mod.rs:558]: Do we need to collect?

Comment (kdy1) [reply crates/swc_ecma_transforms_base/src/resolver/mod.rs:558]: We are iterating right after collecting, so `collect` may not worth

File: crates/swc_ecma_transforms_base/src/resolver/mod.rs [hunk 2]
```diff
@@ -716,6 +728,21 @@ impl<'a> VisitMut for Resolver<'a> {
...
+            {
+                let params = c
+                    .params
+                    .iter()
+                    .filter(|p| match p {
+                        ParamOrTsParamProp::TsParamProp(_) => false,
+                        ParamOrTsParamProp::Param(p) => !p.pat.is_rest(),
+                    })
+                    .flat_map(find_pat_ids)
+                    .collect::<Vec<Id>>();
```
Comment (kdy1) [inline crates/swc_ecma_transforms_base/src/resolver/mod.rs:740]: Same here

**Review by kdy1 — CHANGES_REQUESTED** [review]
Thanks!

---

swc-bump: 
 - swc_ecma_transforms_base

File: crates/swc_ecma_transforms_base/src/resolver/mod.rs (hunk 1 above)
Comment (Austaras) [reply crates/swc_ecma_transforms_base/src/resolver/mod.rs:558]: That's how this PR works -- you may refer latter declared parameters in closures, so we must hoist all parameters before we visit them.

Comment (Austaras) [reply crates/swc_ecma_transforms_base/src/resolver/mod.rs:558]: Same as L893 in `visit_mut_function`

Comment (kdy1) [reply crates/swc_ecma_transforms_base/src/resolver/mod.rs:558]: I see, thank you!

**Review by kdy1 — APPROVED** [review]

File: crates/swc_ecma_transforms_base/src/resolver/mod.rs [hunk 3]
```diff
@@ -549,6 +549,18 @@ impl<'a> VisitMut for Resolver<'a> {
 
             let old = child.ident_type;
             child.ident_type = IdentType::Binding;
+            {
+                let params = e
+                    .params
+                    .iter()
+                    .filter(|p| !p.is_rest())
+                    .flat_map(find_pat_ids)
```
Comment (Austaras) [inline crates/swc_ecma_transforms_base/src/resolver/mod.rs:557]: @kdy1 BTW, do you remember why excluding rest parameter is necessary here?

Comment (kdy1) [reply crates/swc_ecma_transforms_base/src/resolver/mod.rs:557]: Not sure. I think I should write more comments...

**Pull request merged**